doc/pages/user/041-astarte_device_sdks.md

```diff
@@ -0,0 +1,590 @@
+necessary.
+
+### Data Persistence and Automatic Retransmission
+
+Astarte Device SDKs allow configuring persitence and reliability policies. In case of connection
+loss data is stored to memory or disk (according to mappings configuration) and they are
+automatically retransmitted as soon as the device is back online.
+
+This feature is not available yet on Elixir, ESP32, Go and Python SDKs and might be not avilable on
+other platforms with constrained resources.
+
+### Smart Properties Sync
+
+Astarte has support for the concept of properties, which are kept synchronized between the server
+and the device.
+
+Thanks to the [Astarte MQTT v1 Protocol](080-mqtt-v1-protocol.html) an incremental approach is
+employed therefore only changed properties are synchronized. This feature is not available yet on 
+Elixir, Go and Python SDKs and might be not avilable on other SDKs with no `session_present` support.
+
+### Data Validation
+
+Astarte Device SDKs take care of data validation before sending any data, hence errors are reported
+locally on the device improving troubleshooting experience.
+
+This feature is not available yet on ESP32 and is WIP on Rust.
+
+## Device Registration
+
+A device must be registered beforehand to obtain its `credentials-secret`.
+While there are some manual options (such as using the [`astartectl`](https://github.com/astarte-platform/astartectl) command
+or using the [`Astarte Dashboard`](https://docs.astarte-platform.org/snapshot/015-astarte_dashboard.html)), 
+almost all Astarte Device SDKs allow to programmatically register a Device. 
+For Go you can use the [astarte_go](https://github.com/astarte-platform/astarte-go) client.
+
+### Device id
+
+Device ids are 128-bit long url-safe base64 strings without padding. They can be deterministic (UUID v5) or random.
+UUID v5 are obtained from a namespace UUID and a payload (a string).
+While all SDKs work with user-provided device ids, some also provide utilities to for UUID generation.
+
+C (ESP32) with an unique hardware ID using device MAC address and other identification bits:
+```c
+// deterministic id
+astarte_err_t astarte_hwid_get_id(&hw_id);
+```
+
+C++ (Qt5): not supported.
+
+Elixir: UUIDv5 can be obtained using the [elixir_uuid library](https://github.com/zyro/elixir-uuid).
+```elixir
+# random id
+device_id = :crypto.strong_rand_bytes(16) |> Base.url_encode64(padding: false)
+
+#deterministic id
+device_id = UUID.uuid5(namespace_uuid, payload, :raw)
+            |> Astarte.Core.Device.encode_device_id()
+```
+
+Go (using the [Astarte Go Client](https://github.com/astarte-platform/astarte-go)):
+```go 
+// Random id
+random_id, err := GenerateRandomAstarteId()
+
+// Namespaced id
+namespaced_id, err := GetNamespacedAstarteDeviceID(namespaceUuid,payload)
+```
+
+Java/Android: 
+```java
+// Namespaced id
+UUID deviceID = nameUUIDFromNamespaceAndString(namespaceUuid, payload);
```

Comment:
In java you can generate device id (UUID v4 or v5 ) with `AstarteDeviceIdUtils`.

`
String randomID = AstarteDeviceIdUtils.generateId();
`
`
String deviceID = AstarteDeviceIdUtils.generateId(namespaceUuid, payload);
`